app/services/hashed_password.rb

```diff
@@ -0,0 +1,75 @@
+# HashedPassword is returned from PasswordHasher.hash.
+# HashedPassword outputs various key hashing related to NIST encryption.
+#
+# Generate and store a 128-bit salt S.
+# Z1, Z2 = scrypt(S, password)   # split 256-bit output into two halves
+# Generate random R.
+# D = KMS_GCM_Encrypt(key=server_secret, plaintext=R) xor Z1
+# E = hash( Z2 + R )
+# F = hash(E)
+# Store F (User.encrypted_password) and D (User.encryption_key) in db
+
+class HashedPassword
```

Comment:
I find this name confusing, when I see `hashed_password` referenced elsewhere, my expectation is that we're passing around a string or some immediate result, not this intermediate object that itself does some hashing. Did you try any other names? `HashDeriver` or something?